File: cpp/day04/main.cpp
#include "md5.h"
#include <string>

int part1(std::string input) {
  int result = 0;

  while (md5(input + std::to_string(++result)).compare(0, 5, "00000") != 0)
    ;

  return result;
}

int part2(std::string input) {
  int result = 0;

  while (md5(input + std::to_string(++result)).compare(0, 6, "000000") != 0)
    ;

  return result;
}

int main() {
  std::string input = "bgvyzdsv";

  std::cout << "Part 1 answer: " << part1(input)
            << "\nPart 2 answer: " << part2(input) << '\n';
}

File: py/day04/main.py
import hashlib

def part1(secret_key, prefix='00000'):
    num = 1

    while True:
        hash_input = secret_key + str(num)
        hash_result = hashlib.md5(hash_input.encode()).hexdigest()
        if hash_result.startswith(prefix):
            return num
        num += 1

def part2(secret_key, prefix='000000'):
    num = 1

    while True:
        hash_input = secret_key + str(num)
        hash_result = hashlib.md5(hash_input.encode()).hexdigest()
        if hash_result.startswith(prefix):
            return num
        num += 1

def main():
    input = 'bgvyzdsv'
    print(part1(input))
    print(part2(input))
    
if __name__ == "__main__":
    main()
